pkg/volume/portworx/portworx.go

```diff
@@ -0,0 +1,388 @@
+	}
+
+	volumeID := pwx.VolumeID
+	fsType := pwx.FSType
+
+	return &portworxVolumeMounter{
+		portworxVolume: &portworxVolume{
+			podUID:          podUID,
+			volName:         spec.Name(),
+			volumeID:        volumeID,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, spec.Name(), plugin.host)),
+		},
+		fsType:      fsType,
+		readOnly:    readOnly,
+		diskMounter: &mount.SafeFormatAndMount{Interface: plugin.host.GetMounter(), Runner: exec.New()}}, nil
+}
+
+func (plugin *portworxVolumePlugin) NewUnmounter(volName string, podUID types.UID) (volume.Unmounter, error) {
+	return plugin.newUnmounterInternal(volName, podUID, &PortworxVolumeUtil{}, plugin.host.GetMounter())
+}
+
+func (plugin *portworxVolumePlugin) newUnmounterInternal(volName string, podUID types.UID, manager portworxManager, mounter mount.Interface) (volume.Unmounter, error) {
+	return &portworxVolumeUnmounter{
+		&portworxVolume{
+			podUID:          podUID,
+			volName:         volName,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, volName, plugin.host)),
+		}}, nil
+}
+
+func (plugin *portworxVolumePlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec, &PortworxVolumeUtil{})
+}
+
+func (plugin *portworxVolumePlugin) newDeleterInternal(spec *volume.Spec, manager portworxManager) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.PortworxVolume == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.PortworxVolume is nil")
+	}
+	return &portworxVolumeDeleter{
+		portworxVolume: &portworxVolume{
+			volName:  spec.Name(),
+			volumeID: spec.PersistentVolume.Spec.PortworxVolume.VolumeID,
+			manager:  manager,
+			plugin:   plugin,
+		}}, nil
+}
+
+func (plugin *portworxVolumePlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	return plugin.newProvisionerInternal(options, &PortworxVolumeUtil{})
+}
+
+func (plugin *portworxVolumePlugin) newProvisionerInternal(options volume.VolumeOptions, manager portworxManager) (volume.Provisioner, error) {
+	return &portworxVolumeProvisioner{
+		portworxVolume: &portworxVolume{
+			manager: manager,
+			plugin:  plugin,
+		},
+		options: options,
+	}, nil
+}
+
+func (plugin *portworxVolumePlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
+	portworxVolume := &v1.Volume{
+		Name: volumeName,
+		VolumeSource: v1.VolumeSource{
+			PortworxVolume: &v1.PortworxVolumeSource{
+				VolumeID: volumeName,
+			},
+		},
+	}
+	return volume.NewSpecFromVolume(portworxVolume), nil
+}
+
+func getVolumeSource(
+	spec *volume.Spec) (*v1.PortworxVolumeSource, bool, error) {
+	if spec.Volume != nil && spec.Volume.PortworxVolume != nil {
+		return spec.Volume.PortworxVolume, spec.ReadOnly, nil
```

Comment:
`spec.ReadOnly` is not trusted in this case. You need to provide readonly in portworx volume source.